lib/pbench/server/templates.py

```diff
@@ -256,13 +146,276 @@ def __init__(self, basepath, idx_prefix, logger, known_tool_handlers=None, _dbg=
+                "mappings" or "properties" keys.
+        """
+        self.mappings.add_mapping(name, body)
+
+    def resolve(self):
+        """
+        Expand templates to create all the information necessary to register a
+        document template with Elasticsearch.
+
+        Some templates share settings JSON files, and the tool templates are
+        generated from a skeleton/base and tool "fragment"; all of this logic
+        is managed here.
+
+        Once this method returns, the body() method can be called to return
+        the full Elasticsearch JSON payload necessary to register a document
+        template and its index pattern.
+        """
+        self.load()
+
+        # FIXME: It's ugly reaching into the JsonFile object from here; this
+        # ought to be refactored into JsonFile somehow.
+        if self.tool:
+            tool_mapping = copy.deepcopy(self.base.mappings.json)
+
+            # This is ugly: we get the version from the tool-specific overlay;
+            # but that becomes a subdocument of the skeleton, and we need the
+            # version there. So we delete from one and create in the other.
+            del self.mappings.json["_meta"]
+            tool_mapping["_meta"] = dict(version=self.version)
+            sub_map = self.mappings.json
+            self.mappings.json = tool_mapping
+            self.add_mapping(self.name, sub_map)
+
+        # Add a standard "authorization" sub-document into the document
+        # template if this document type is "owned" by a Pbench user.
+        if self.index_info["owned"]:
+            self.add_mapping(
+                "authorization",
+                {
+                    "properties": {
+                        "owner": {"type": "keyword"},
+                        "access": {"type": "keyword"},
+                    }
+                },
+            )
+
+    def body(self) -> Dict[AnyStr, Any]:
+        """
+        Return a JSON payload suitable to POST to Elasticsearch in order to
+        register an indexed document template.
+
+        Returns:
+            Elasticsearch template payload
+        """
+        return dict(
+            index_patterns=self.index_pattern,
+            settings=self.settings.json,
+            mappings=self.mappings.json,
+        )
+
+
+class PbenchTemplates:
+    """Encapsulation of methods for loading / working with all the Pbench
+    templates for Elasticsearch.
+    """
+
+    def __init__(self, basepath, idx_prefix, logger, known_tool_handlers=None, _dbg=0):
+        # Determine the location of the mapping and settings files. We expect
+        # the "basepath" parameter to be the PbenchServerConfig BINDIR attr,
+        # which will be something like /opt/pbench-server/bin
+        base_dir = Path(basepath).parent / "lib"
+        mapping_dir = base_dir / "mappings"
+        # Where to find the settings
+        setting_dir = base_dir / "settings"
```

Comment:
This is difficult to parse since your line numbers don't match the code against which GitHub displays it. I think that, at the time, I thought that the `config.BASEPATH` reference was useful, and after explaining that, failing to explain the equivalent `setting_dir` felt awkward. However since this is a parameter and not a direct reference to the config object, you're probably right that it's all unnecessary.

Reply:
The line numbers look OK if you read this in the "Conversation" tab.  :-)

(They would probably look OK, too, reading it in "Files changed" tab if you had the right commit selected....)